File: ui/scripts/find_unused_translations.py
#!/usr/bin/env python3

import re
import subprocess
import sys
import argparse
from pathlib import Path

# Configuration
LOCALES_PATH = "src/locales"
UI_SRC_PATH = "src"
LOCALE_TO_CHECK = "en.ts"

# ANSI color codes for output formatting
GREEN = "\033[92m"
YELLOW = "\033[93m"
RED = "\033[91m"
RESET = "\033[0m"


def extract_translation_keys(file_path):
    """Extract all translation keys from the locale file."""
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()

    # Pattern to match translation keys: key: 'value' or key: (params) => `...`
    pattern = r"^\s*([a-zA-Z_][a-zA-Z0-9_]*)\s*:"

    keys = []
    for line in content.splitlines():
        match = re.match(pattern, line)
        if match:
            key = match.group(1)
            keys.append(key)

    return keys


def is_key_used(key, src_path):
    """Check if a translation key is used in the codebase."""
    # Different patterns to search for
    patterns = [
        f"t\\('{key}'\\)",
        f't\\("{key}"\\)',
        f"t\\(`{key}`\\)",
        f"'{key}'",
        f'"{key}"',
        f"`{key}`",
    ]

    for pattern in patterns:
        try:
            cmd = [
                "grep",
                "-r",
                pattern,
                str(src_path),
                "--include=*.tsx",
                "--include=*.ts",
            ]

            # Run grep command and filter out matches in locales directory
            result = subprocess.run(cmd, capture_output=True, text=True, check=False)

            if result.returncode == 0:
                # Filter out matches in locale files
                filtered_lines = [
                    line
                    for line in result.stdout.splitlines()
                    if "locales/" not in line
                ]

                if filtered_lines:
                    return True
        except subprocess.SubprocessError:
            print(f"Error running grep for key '{key}'")

    return False


def main():
    """Find unused translation keys in the locale file."""
    parser = argparse.ArgumentParser(
        description="Find unused translation keys in locale files"
    )
    parser.add_argument(
        "--ci",
        action="store_true",
        help="Run in CI mode (fails on unused translations)",
    )
    parser.add_argument("--quiet", action="store_true", help="Reduce output verbosity")
    args = parser.parse_args()

    ci_mode = args.ci
    quiet_mode = args.quiet

    if not quiet_mode:
        print(f"Checking for unused translation keys in {LOCALE_TO_CHECK}...")

    locale_file_path = Path(f"{LOCALES_PATH}/{LOCALE_TO_CHECK}")
    if not locale_file_path.exists():
        print(f"Error: Locale file {LOCALE_TO_CHECK} not found!")
        sys.exit(1)

    keys = extract_translation_keys(locale_file_path)
    if not quiet_mode:
        print(f"Found {len(keys)} translation keys in {LOCALE_TO_CHECK}")

    unused_keys = []
    used_keys = []

    # Process each key
    for key in keys:
        if not quiet_mode:
            print(f"Checking key: {key}... ", end="")

        if not is_key_used(key, UI_SRC_PATH):
            if not quiet_mode:
                print(f"{RED}UNUSED{RESET}")
            unused_keys.append(key)
        else:
            if not quiet_mode:
                print(f"{GREEN}USED{RESET}")
            used_keys.append(key)

    # Print summary
    print("\n--- SUMMARY ---")
    if not unused_keys:
        print(f"{GREEN}No unused translation keys found!{RESET}")
    else:
        print(
            f"{RED if ci_mode else YELLOW}Found {len(unused_keys)} unused translation keys:{RESET}"
        )
        for key in unused_keys:
            print(f"- {key}")

    print(f"\nTotal keys: {len(keys)}")
    print(f"Used keys: {len(used_keys)}")
    print(f"Unused keys: {len(unused_keys)}")

    # In CI mode, exit with error if unused keys are found
    if ci_mode and unused_keys:
        print(
            f"{RED}Error: Found {len(unused_keys)} unused translation keys in CI mode!{RESET}"
        )
        sys.exit(1)


if __name__ == "__main__":
    main()

File: ui/scripts/find_unused_translations.js
#!/usr/bin/env node
/*  find-unused-translations.js
    --------------------------------------------------------------
    Reports (or in CI mode, fails on) translation keys that exist
    in the default locale file but are never referenced in the UI
    source tree.                                                   */

import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'

// ─── Configuration ──────────────────────────────────────────────
const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

const LOCALES_PATH = path.join(__dirname, '../src/locales')
const UI_SRC_PATH = path.join(__dirname, '../src')
const LOCALE_TO_CHECK = 'en.ts' // change if you need another default

// ─── ANSI helpers (pretty printing) ─────────────────────────────
const GREEN = '\x1b[32m'
const YELLOW = '\x1b[33m'
const RED = '\x1b[31m'
const RESET = '\x1b[0m'
// ────────────────────────────────────────────────────────────────

// ─── CLI flags ─────────────────────────────────────────────────
const args = process.argv.slice(2)
const ciMode = args.includes('--ci')
const quietMode = args.includes('--quiet')
// ────────────────────────────────────────────────────────────────

// ─── Step 1: get all translation keys ──────────────────────────
function extractKeys(localePath) {
  const src = fs.readFileSync(localePath, 'utf-8')
  const pattern = /^\s*([a-zA-Z_][a-zA-Z0-9_]*)\s*:/gm
  const keys = []
  let m
  while ((m = pattern.exec(src)) !== null) keys.push(m[1])
  return keys
}

const localeFile = path.join(LOCALES_PATH, LOCALE_TO_CHECK)
if (!fs.existsSync(localeFile)) {
  console.error(`❌  Locale file '${LOCALE_TO_CHECK}' not found`)
  process.exit(1)
}
const keys = extractKeys(localeFile)
if (!quietMode)
  console.log(`🔍  Found ${keys.length} keys in ${LOCALE_TO_CHECK}`)
// ────────────────────────────────────────────────────────────────

// ─── Step 2: load all .ts/.tsx source files into memory ────────
function walk(dir) {
  const items = fs.readdirSync(dir, { withFileTypes: true })
  const files = []
  for (const item of items) {
    const full = path.join(dir, item.name)
    if (item.isDirectory()) {
      if (full.includes(`${path.sep}locales${path.sep}`)) continue // skip locale files
      files.push(...walk(full))
    } else if (item.name.endsWith('.ts') || item.name.endsWith('.tsx')) {
      files.push(full)
    }
  }
  return files
}
const sourceFiles = walk(UI_SRC_PATH)
const fileContents = sourceFiles.map((p) => fs.readFileSync(p, 'utf-8'))
// ────────────────────────────────────────────────────────────────

// ─── Step 3: helper to test if a key is referenced ─────────────
const escapeRegExp = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
function buildRegexes(key) {
  const k = escapeRegExp(key)
  return [
    new RegExp(`t\\('${k}'\\)`),
    new RegExp(`t\\("${k}"\\)`),
    new RegExp(`t\\(\`${k}\`\\)`),
    new RegExp(`'${k}'`),
    new RegExp(`"${k}"`),
    new RegExp(`\`${k}\``),
  ]
}
function keyIsUsed(key) {
  const rex = buildRegexes(key)
  for (const content of fileContents) {
    if (rex.some((r) => r.test(content))) return true
  }
  return false
}
// ────────────────────────────────────────────────────────────────

// ─── Step 4: iterate over keys ─────────────────────────────────
const unused = []
const used = []

for (const key of keys) {
  if (!quietMode) process.stdout.write(` • ${key.padEnd(40)} `)
  if (keyIsUsed(key)) {
    used.push(key)
    if (!quietMode) console.log(`${GREEN}USED${RESET}`)
  } else {
    unused.push(key)
    if (!quietMode) console.log(`${RED}UNUSED${RESET}`)
  }
}
// ────────────────────────────────────────────────────────────────

// ─── Final summary ─────────────────────────────────────────────
console.log('\n— SUMMARY —')
if (unused.length === 0) {
  console.log(`${GREEN}No unused translation keys found!${RESET}`)
} else {
  const colour = ciMode ? RED : YELLOW
  console.log(
    `${colour}Found ${unused.length} unused translation key(s):${RESET}`
  )
  unused.forEach((k) => console.log(' -', k))
}

console.log(`\nTotal keys : ${keys.length}`)
console.log(`Used       : ${used.length}`)
console.log(`Unused     : ${unused.length}`)

// ─── CI-mode exit code ─────────────────────────────────────────
if (ciMode && unused.length) {
  console.error(
    `${RED}\nCI mode: failing build because of unused keys.${RESET}`
  )
  process.exit(1)
}
